<launch>
	<!-- include -->
	<include file="$(find duckietown)/machines"/>
	<arg name="verbose" default="false" />
	<arg name="joystick" default="true" doc="true for listening to the joystick. False for listening to the coordination."/>
	<arg name="veh" doc="Name of vehicle. ex: megaman"/>
	<arg name="local" default="false" doc="true for running everything you possibly can on laptop. false for everything running on vehicle."/>
	<arg name="config" default="baseline" doc="Specify a config."/>
	<arg name="param_file_name" default="default" doc="Specify a param file. ex:megaman." />
	<arg name="continuous" default="True"/>
	<arg name="intersectionType" default="trafficLight" />

	<group ns="$(arg veh)">
		<!-- Run local -->

		<!-- Run coordination node -->
		<node if="$(arg local)" name="coordinator_node" pkg="explicit_coordinator" type="coordinator_node_ETHZ17_light.py" output="screen" clear_params="true" required="true">
			<!-- <rosparam command="load" file="$(find duckietown)/config/$(arg config)/explicit_coordination/veh_coordinator/simple_coordinator_node/$(arg param_file_name).yaml"/> -->
			<param name="verbose" value="false"/>
			<param name="intersectionType" value="$(arg intersectionType)"/>
			<!-- Input to coordination node -->
			<remap from="~signals_detection" to="led_interpreter_node/signals_detection"/>
		</node>

		<!-- Run with detection node-->
		<node if="$(arg local)" pkg="led_detection" type="LED_detector_node.py" name="led_detector_node" output="screen" clear_params="true" required="true">
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/LED_protocol.yaml"/>
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_detection/LED_detector_node/$(arg param_file_name).yaml"/>
			<param name="continuous" type="bool" value="$(arg continuous)" />
		</node>

		<!-- Run with interpreter node-->
		<node if="$(arg local)"  pkg="led_interpreter" type="LED_interpreter_node.py" name="led_interpreter_node" output="screen" clear_params="true" required="true">
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/LED_protocol.yaml"/>
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/location_config.yaml"/>
			<param name="continuous" type="bool" value="$(arg continuous)" />
                        <param name="intersectionType" value="$(arg intersectionType)" />
			<remap from="led_interpreter_node/raw_led_detection" to="led_detector_node/raw_led_detection" />
		</node>

		<!-- Run with emitter node -->
		<node if="$(arg local)" pkg="led_emitter" type="led_emitter_node.py"  name="led_emitter_node" output="screen" clear_params="true" required="true">
			<!-- other -->
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/LED_protocol.yaml"/>
			<!-- Input from coordination node -->
			<remap from="~change_color_pattern" to="coordinator_node/change_color_pattern"/>
		</node>



	    <!-- run on vehicle -->

	    <!-- Run coordination node -->
		<node unless="$(arg local)" machine="$(arg veh)" name="coordinator_node" pkg="explicit_coordinator" type="coordinator_node_ETHZ17_light.py" output="screen" clear_params="true" required="true">
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/veh_coordinator/simple_coordinator_node/$(arg param_file_name).yaml"/>
			<param name="verbose" value="false"/>
			<param name="intersectionType" value="$(arg intersectionType)"/>
			<!-- Input to coordination node -->
			<remap from="~signals_detection" to="led_interpreter_node/signals_detection"/>
		</node>

		<!-- Run with detection node-->
		<node unless="$(arg local)"  machine="$(arg veh)" pkg="led_detection" type="LED_detector_node.py" name="led_detector_node" output="screen" clear_params="true" required="true">
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/LED_protocol.yaml"/>
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_detection/LED_detector_node/$(arg param_file_name).yaml"/>
			<param name="continuous" type="bool" value="$(arg continuous)" />
		</node>

		<!-- Run with interpreter node-->
		<node unless="$(arg local)" machine="$(arg veh)" pkg="led_interpreter" type="LED_interpreter_node.py" name="led_interpreter_node" output="screen" clear_params="true" required="true">
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/LED_protocol.yaml"/>
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/location_config.yaml"/>
			<param name="continuous" type="bool" value="$(arg continuous)" />
			<param name="intersectionType" value="$(arg intersectionType)" />
			<remap from="led_interpreter_node/raw_led_detection" to="led_detector_node/raw_led_detection" />
		</node>

		<!-- Run with emitter node -->
		<node unless="$(arg local)" machine="$(arg veh)" pkg="led_emitter" type="led_emitter_node.py"  name="led_emitter_node" output="screen" clear_params="true" required="true">
			<!-- other -->
			<rosparam command="load" file="$(find duckietown)/config/$(arg config)/led_interpreter/LED_protocol.yaml"/>
			<!-- Input from coordination node -->
			<remap from="~change_color_pattern" to="coordinator_node/change_color_pattern"/>
		</node>

	</group>


    <!-- Publications -->

    <!-- Subscriptions -->

</launch>
